woocommerce/woocommerce
# Allow filtering variations by attributes in API.

### All Submissions:






### Changes proposed in this Pull Request:



Closes #20349 .

### How to test the changes in this Pull Request:

1. Create attribute, e.g. color.
2. Create terms, e.g. yellow, blue, get ids of those terms.
3. Create shipping classes, e.g. `sc_1` and `sc_2`, get ids of those.
4. Create a tag, e.g. `tag_1`, get the id of the tag.
5. Create a category, e.g. `cat_1`, get the id of the category.
6.  Create a variable product with at least 3 variations: e.g. color = blue, color = yellow, color = unspecified
7. Set shipping class `sc_1` on the product level and assign `sc_2` to one of the variations.
8. Assign tag `tag_1` to the variable product.
9. Assign category `cat_1` to the variable product.
10. Set the product as featured.
11. Test `/wc/v2/products/{id}`, observe that `categories` and `tags` is empty.
12. Test `wc/v2/products/{id}/variations/?category=cat_1_id`, result is empty set.
13. Test `wc/v2/products/{id}/variations/?type=variable`, result is empty set.
14. Test `wc/v2/products/{id}/variations/?featured=true`, result is empty set. 
15. Test `wc/v2/products/{id}/variations/?featured=false`, it returns all variations, despite the product being featured.
16.  Test `wc/v2/products/{id}/variations/?tag=tag_1_id`, result is empty set.
17.  Test `wc/v2/products/{id}/variations/?shipping_class=sc_1_id`, result is empty set.
18.  Test `wc/v2/products/{id}/variations/?shipping_class=sc_2_id`, result is correct this time, as shipping class is assigned to the variation.
19. Test `wc/v2/products/{id}/variations/?attribute=pa_color&attribute_term=yellow_id`, result is empty set.
20. Apply branch
11. Test `/wc/v2/products/{id}`, observe that `categories` and `tags` is not empty anymore.
12. Test `wc/v2/products/{id}/variations/?category=cat_1_id`, result is array of all variations.
13. Test `wc/v2/products/{id}/variations/?type=variable`, result is array of all variations.
14. Test `wc/v2/products/{id}/variations/?featured=true`, result is array of all variations. 
15. Test `wc/v2/products/{id}/variations/?featured=false`, result is empty set, thus correct..
16.  Test `wc/v2/products/{id}/variations/?tag=tag_1_id`, result is array of all variations.
17.  Test `wc/v2/products/{id}/variations/?shipping_class=sc_1_id`, result is array of all variations that have shipping class 1 assigned to them.
18.  Test `wc/v2/products/{id}/variations/?shipping_class=sc_2_id`, result is array of all variations that have shipping class 1 assigned to them.
19. Test `wc/v2/products/{id}/variations/?attribute=pa_color&attribute_term=yellow_id`, result is array of all variations that have attribute color set to yellow.

### Other information:




### Changelog entry

> Allow filtering variations by attributes via query parameters in REST API.
